File: app/src/main/java/com/OdiousPanda/thefweather/Utilities/UnitConverter.java
package com.OdiousPanda.thefweather.Utilities;

import java.util.Random;

public class UnitConverter {

    private final static String DEGREE  = "\u00b0";

    public static float toCelsius(float temp){
        return (temp - 32)* 5/9;
    }

    private static float toFahrenheit(float temp){
        return temp;
    }

    private static String toCelsiusPretty(float temp){
        return Math.round(toCelsius(temp)) + DEGREE + "C";
    }

    private static String toFahrenheitPretty(float temp){
        return Math.round(toFahrenheit(temp)) + DEGREE + "F";
    }
    private static float toKelvin(float temp){
        return toCelsius(temp) + 273;
    }

    private static String toKelvinPretty(float temp){
        return Math.round(toKelvin(temp)) + DEGREE + "K";
    }

    public static String convertToTemperatureUnit(float temp, String unit){
        if(unit.equals(DEGREE + "C")){
            return toCelsiusPretty(temp);
        }
        else if(unit.equals(DEGREE + "F")){
            return toFahrenheitPretty(temp);
        }
        return toKelvinPretty(temp);
    }

    public static String convertToDistanceUnit(float distance, String unit){
        if(unit.equals("km")){
            return Math.round(distance * 1.609) + " km";
        }
        else if(unit.equals("bananas")){
            return Math.round(distance * 9041.254) + " bananas";
        }
        return Math.round(distance) + " mile";
    }

    public static String convertToSpeedUnit(float speed, String unit){
        if(unit.equals("kmph")){
            return Math.round(speed * 1.609) + " km/h";
        }
        else if(unit.equals("bananas per hour")){
            return Math.round(speed * 9041.254) + " banana/h";
        }
        return Math.round(speed) + " mph";
    }

    public static float toMeterPerSecond(float speed){
        return (float)(speed * 0.447);
    }

    public static String convertToPressureUnit(float pressure, String unit){
        if(unit.equals("psi")){
            return Math.round(pressure / 68.948) + " psi";
        }
        else if(unit.equals("mmHg")){
            return Math.round(pressure / 1.333) + " mmHg";
        }
        else{
            String[] depressLevels = {"Mild","Depressed","Crippling","A little","Let's die","Oof","Happy af!","Not at all","Over 9000"};
            return depressLevels[new Random().nextInt(depressLevels.length)];
        }
    }
}

File: app/src/main/java/com/odiousPanda/rainbowKt/utilities/UnitConverter.kt
package com.odiousPanda.rainbowKt.utilities

import android.content.Context
import com.odiousPanda.rainbowKt.R
import java.nio.charset.StandardCharsets
import java.util.*
import kotlin.math.roundToInt

object UnitConverter {
    private const val DEGREE = "\u00b0"

    private fun banana(): String {
        val bananaByte = byteArrayOf(
            0xF0.toByte(),
            0x9F.toByte(),
            0x8D.toByte(),
            0x8C.toByte()
        )
        return String(bananaByte, StandardCharsets.UTF_8)
    }

    fun toCelsius(temp: Float): Float {
        return (temp - 32) * 5 / 9
    }

    private fun toFahrenheit(temp: Float): Float {
        return temp
    }

    private fun toCelsiusPretty(temp: Float): String {
        return "${toCelsius(temp).roundToInt()}${DEGREE}C"
    }

    private fun toFahrenheitPretty(temp: Float): String {
        return "${toFahrenheit(temp).roundToInt()}${DEGREE}F"
    }

    private fun toKelvin(temp: Float): Float {
        return toCelsius(temp) + 273
    }

    private fun toKelvinPretty(temp: Float): String {
        return "${toKelvin(temp).roundToInt()}$DEGREE}K"
    }

    fun convertToTemperatureUnit(temp: Float, unit: String): String {
        if (unit == "${DEGREE}C") {
            return toCelsiusPretty(temp)
        } else if (unit == "${DEGREE}F") {
            return toFahrenheitPretty(temp)
        }
        return toKelvinPretty(temp)
    }

    fun convertToTemperatureUnitClean(
        temp: Float,
        unit: String
    ): String? {
        if (unit == "${DEGREE}C") {
            return toCelsius(temp).roundToInt().toString() + DEGREE
        } else if (unit == "${DEGREE}F") {
            return toFahrenheit(temp).roundToInt().toString() + DEGREE
        }
        return "${toKelvin(temp).roundToInt()}${DEGREE}K"
    }

    fun convertToDistanceUnit(distance: Float, unit: String): String {
        if (unit == "km") {
            return "${(distance * 1.609).roundToInt()} km"
        } else if (unit == "mile") {
            return "${distance.roundToInt()} mi"
        }
        return "${(distance * 9041.254).roundToInt()} ${banana()}"
    }

    fun convertToSpeedUnit(speed: Float, unit: String): String {
        if (unit == "kmph") {
            return "${(speed * 1.609).roundToInt()} km/h"
        } else if (unit == "mph") {
            return "${speed.roundToInt()} mph"
        }
        return "${(speed * 9041.254).roundToInt()} ${banana()}/h"
    }

    fun toMeterPerSecond(speed: Float): Float {
        return (speed * 0.447).toFloat()
    }

    fun convertToPressureUnit(
        context: Context,
        pressure: Float,
        unit: String
    ): String? {
        return when (unit) {
            "psi" -> {
                "${(pressure / 68.948).roundToInt()} psi"
            }
            "mmHg" -> {
                "${(pressure / 1.333).roundToInt()} mmHg"
            }
            else -> {
                val depressLevels =
                    context.resources.getStringArray(R.array.depression_levels)
                depressLevels[Random().nextInt(depressLevels.size)]
            }
        }
    }
}